From 22339e52da4e88af499e232e452f4d8bb358fa05 Mon Sep 17 00:00:00 2001
From: Yabin Cui <yabinc@google.com>
Date: Mon, 14 Nov 2022 17:48:44 +0000
Subject: [PATCH] [Support] Use thread safe version of getpwuid and getpwnam.

OpenGroup specification doesn't require getpwuid and getpwnam
to be thread-safe. And musl libc has a not thread-safe implementation.
When building clang with musl, this can make clang-scan-deps crash.

Reviewed By: pirama

Differential Revision: https://reviews.llvm.org/D137864
---
 llvm/lib/Support/Unix/Path.inc | 21 +++++++++++++++++----
 1 file changed, 17 insertions(+), 4 deletions(-)

diff --git a/llvm/lib/Support/Unix/Path.inc b/llvm/lib/Support/Unix/Path.inc
index 3ecc5c971b79..2eb17474c46f 100644
--- a/llvm/lib/Support/Unix/Path.inc
+++ b/llvm/lib/Support/Unix/Path.inc
@@ -676,11 +676,17 @@ static void expandTildeExpr(SmallVectorImpl<char> &Path) {
 
   // This is a string of the form ~username/, look up this user's entry in the
   // password database.
-  struct passwd *Entry = nullptr;
+  std::unique_ptr<char[]> Buf;
+  long BufSize = sysconf(_SC_GETPW_R_SIZE_MAX);
+  if (BufSize <= 0)
+    BufSize = 16384;
+  Buf = std::make_unique<char[]>(BufSize);
+  struct passwd Pwd;
   std::string User = Expr.str();
-  Entry = ::getpwnam(User.c_str());
+  struct passwd *Entry = nullptr;
+  getpwnam_r(User.c_str(), &Pwd, Buf.get(), BufSize, &Entry);
 
-  if (!Entry) {
+  if (!Entry || !Entry->pw_dir) {
     // Unable to look up the entry, just return back the original path.
     return;
   }
@@ -1339,9 +1345,16 @@ std::error_code changeFileOwnership(int FD, uint32_t Owner, uint32_t Group) {
 namespace path {
 
 bool home_directory(SmallVectorImpl<char> &result) {
+  std::unique_ptr<char[]> Buf;
   char *RequestedDir = getenv("HOME");
   if (!RequestedDir) {
-    struct passwd *pw = getpwuid(getuid());
+    long BufSize = sysconf(_SC_GETPW_R_SIZE_MAX);
+    if (BufSize <= 0)
+      BufSize = 16384;
+    Buf = std::make_unique<char[]>(BufSize);
+    struct passwd Pwd;
+    struct passwd *pw = nullptr;
+    getpwuid_r(getuid(), &Pwd, Buf.get(), BufSize, &pw);
     if (pw && pw->pw_dir)
       RequestedDir = pw->pw_dir;
   }
-- 
2.38.1.493.g58b659f92b-goog

